qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -131,82 +132,37 @@ def __getitem__(self, item: Union[int, str]):
             return self._x[item]
```

Comment:
I wrote the docstring and felt that the implementation is a bit tricky. I think it would be simpler to remove `__getitem__` to keep `OptimizationResult` simpler. Users can access the value of variables by `OptimizationResult.x[i]` if `i` is an integer and `OptimizationResult.variables_dict[name]` is `name` is a string. I remember that @adekusar-drl suggested this method. Do you want to keep `__getitem__`?

Reply:
@t-imamichi I think get item is a very convenient method. If there's no obvious drawback of difficulties, it's better to have it.

Reply:
I have no issue with keeping method it does seem useful in some ways if you know about this behavior. Hence above when I gave the examples result[0] and result['A'] and referred to main docstring I was more thinking of some description in the class docstring that described this behavior rather than a docstring around the params etc on the getitem method. I do not know how the latter comes out, if at all, in the API docs. Its fine to have that on the method too but my original thought was more to document it in place where the user would more come across this. I did not pull this branch to check the html outcome as it stands.

Reply:
OK. I keep `__getitem__`. I will add an example of the main docstrings of `OptimizationResult`. How about implementing `__len__`? It might be convenient to use with `__getitem__` with index.

Reply:
Can I step back and ask some question. The optimal value is a list - how do I know the order in this list? Is it a one to one correspondence to the index of the variable that the quadratic program gives me. It seems the values are stored separately from the variables rather than being some field with the variables. Now the variables seems to be an Optional param for the result which means I need to know how the optimization result was built by the particular optimizer which may or may not provide the variables to the result. If it does not then the dictionary and names list I can get back are empty. It seems I can get back the optimal value as a list, or directly an item from that off the result via the get item, and if it were given variables then via a variable name on get item or a dictionary. There seem to be a few choices here, which maybe is not all bad, but then if variables are not provided I am left with the basic by index on the values only. If that is all I can rely on, knowing say the index from the Quadratic program variable to access the result why would I code it any other way? If I try to use the dictionary or access by name it will fail if variables were not supplied to the result so my code would have to be written that tried that and then fell back to by index. Given that's extra work, and the value is unclear to me since it has to cater to the basic case I guess I am asking about the overall usability of result here since things are being changed around it. Could variables me made mandatory to make the overall result more consistent in its behavior? If not I guess I am wondering why we have all that extra stuff based on names when I most likely am forced to really code by index to have it work everywhere.

Reply:
1. Variables/solution values must be indexed in the same way. Originally variables are stored as a list in `QuadraticProgram`, so in `OptimizationResult` solution values must be in the same order if you want access them by index.
2. I thought we decided to have variables (or at least variable names) as a mandatory parameter in  `OptimizationResult`. It does not cost a thing, since we already have a list of variables in `QudraticProgram`.
3. `__getitem__` is for convenience only, sometimes when you code a problem you want to solve, you remember variable names, but don't remember which of them was added first or second. It is similar to the behavior we've already introduced: `qp.objective.linear[index]`. Yes, a different story but in the code looks similar.
4. For `__len__` I don't see a real need at least now.

Reply:
For 1. above the responsibility of the ordering is the optimizer producing the result, right? I did not see any definition of ordering hence I assumed they were the same as the QP as any other order did not seem to make sense.

For 2. certainly having it as mandatory would make the result more consistent/user friendly. Without it being mandatory you just can't rely on it at all, hence my comments.

For 3. get item by name (if you don't know the index) only works when it has the variables (Optional at present). There is also the dictionary return of key/values, when variables is provided, so I can do result['A'] or result.variables_dict['A'], the latter using the dictionary the former the result directly. Maybe its useful to have the whole dict be passed back as an option - either way access by name requires the variables. Part of my question around ordering was if I needed the index I guessed I could look it up in the QP, but that assumed the index order was the same which I did not see explicitly stated anywhere.

Reply:
For 1. Both optimizers and converters are responsible for the ordering of the variables of the input QP. But, it is not explicitly stated. For example, `IntegerToBinary` and `IntegerToBinary` change come variables, so we need to take care of the ordering of variables of `encode` and `decode` (`convert` and `interpret`, new names).

It might be better to explicitly state the ordering of variables of `OptimizationResult` in `OptimizationAlgorithm` and `QuadraticProgramConverter` of #1061.

For 2. It is OK to make `variables` mandatory. It makes the code simpler.

Reply:
I updated the docstring to add examples. Please take a look at it.

Reply:
I agree with @t-imamichi. The initial order of the variables comes from `QuadraticProgram`, after that it must be maintained in the rest code. Not great that it is not explicitly stated though.

Reply:
I try to state it in the main docstring.

Reply:
I updated the main docstring with examples.